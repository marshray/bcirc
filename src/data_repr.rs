#![allow(dead_code)] //? TODO for development
#![allow(unused_mut)] //? TODO for development
#![allow(unused_variables)] //? TODO for development
#![allow(unused_imports)] //? TODO for development

use anyhow::*;
use serde::{Deserialize, Serialize};
//use thiserror::Error;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum IntegerRepr {
    I128(i128)
}
