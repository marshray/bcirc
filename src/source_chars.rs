// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

#![allow(dead_code)] //? TODO for development
#![allow(unused_mut)] //? TODO for development
#![allow(unused_variables)] //? TODO for development
#![allow(unused_imports)] //? TODO for development

use std::ops::{Generator, GeneratorState, RangeInclusive};
use std::pin::Pin;

use anyhow::*;
use serde::{Deserialize, Serialize};

use crate::maybe_match;
use crate::source_bytes::{ByteOrEof, SourceByteReadResult};

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
pub struct LineNumberOneBased(u64);
impl LineNumberOneBased {
    pub fn inc(&mut self) {
        self.0 += 1;
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
pub struct CharNumberOneBased(u64);

impl CharNumberOneBased {
    pub fn inc(&mut self) {
        self.0 += 1;
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct CharLoc {
    pub file_offset: RangeInclusive<u64>,
    pub line_n: LineNumberOneBased,
    pub char_n: CharNumberOneBased,
}

impl CharLoc {
    pub fn new() -> CharLoc {
        CharLoc {
            file_offset: 0u64..=0,
            line_n: LineNumberOneBased(1),
            char_n: CharNumberOneBased(1),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum CharError {
    InvalidUtf8(u8),
    UnexpectedEof,
    StdIoError(String),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum CharResult {
    Char(char),
    Eol,
    Eof,
    CharError(CharError),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SourceCharReadResult {
    pub loc: CharLoc,
    pub char_result: CharResult,
}

impl SourceCharReadResult {
    pub fn default_error() -> Self {
        Self {
            loc: CharLoc::new(),
            char_result: CharResult::Eof,
        }
    }
}

impl SourceCharReadResult {
    /// Returns if the result is a valid Eol.
    pub fn is_eol(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Eol,
                ..
            }
        )
    }

    /// Returns if the result is a valid Eof.
    pub fn is_eof(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Eof,
                ..
            }
        )
    }

    /// Returns if the result is an error.
    pub fn is_error(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::CharError(_),
                ..
            }
        )
    }

    /// Returns if the result is a char.
    pub fn is_char(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Char(ch),
                ..
            }
        )
    }

    /// Returns if the result is a char or a valid Eol.
    pub fn is_char_or_eol(&self) -> bool {
        self.is_char() || self.is_eol()
    }

    /// Returns the char, if the SourceCharReadResult contains one.
    pub fn opt_char(&self) -> Option<char> {
        maybe_match!(*self, SourceCharReadResult { char_result: CharResult::Char(ch), .. } => ch)
    }

    /// Returns the char if the SourceCharReadResult contains one. Otherwise returns
    /// NUL (0 as char).
    pub fn char_or_nul(&self) -> char {
        self.opt_char().unwrap_or('\0')
    }
}

pub fn source_chars(
    source_bytes: &mut dyn Iterator<Item = SourceByteReadResult>,
) -> impl Iterator<Item = SourceCharReadResult>  + '_ {
    let mut source_char_read_result = SourceCharReadResult {
        loc: CharLoc::new(),
        char_result: CharResult::Eof,
    };

    //let mut file_offset_buf0 = 0u64;
    let mut utf8_buf = [0u8; 4];
    let mut utf8_buf_len = 0;
    let mut need_more_bytes_for_this_char = false;
    let mut track_prev_char_was_cr = false;

    //let mut generate_source_chars = move
    std::iter::from_generator(move || {
        for source_byte_read_result in source_bytes {
            source_char_read_result.loc.file_offset = if need_more_bytes_for_this_char {
                // Need more bytes for the current char
                *source_char_read_result.loc.file_offset.start()
            } else {
                // Begin a new char
                utf8_buf_len = 0;
                source_byte_read_result.file_offset
            }..=source_byte_read_result.file_offset;

            need_more_bytes_for_this_char = false;

            match source_byte_read_result.byte_or_eof {
                ByteOrEof::Byte(by) => {
                    assert!(utf8_buf_len < 4);
                    utf8_buf[utf8_buf_len] = by;
                    utf8_buf_len += 1;

                    match std::str::from_utf8(&utf8_buf[0..utf8_buf_len]) {
                        Result::Ok(ch_str) => {
                            assert_eq!(ch_str.len(), 1);
                            let ch = ch_str.chars().next().unwrap();

                            let prev_char_was_cr = track_prev_char_was_cr;
                            track_prev_char_was_cr = ch == '\r';

                            if prev_char_was_cr && ch == '\n' {
                                // Ignore LF after CR
                            } else {
                                let (yielding_eol, char_or_something) = if ch == '\r' || ch == '\n' {
                                    (true, CharResult::Eol)
                                } else {
                                    (false, CharResult::Char(ch))
                                };

                                let _char_or_eof = char_or_something.clone();
                                let scrr = SourceCharReadResult {
                                    char_result: char_or_something,
                                    ..source_char_read_result.clone()
                                };
                                //eprintln!("\nyielding {:?}", &scrr);

                                yield scrr;

                                if yielding_eol {
                                    //eprintln!("advancing to next line: {:?}", &_char_or_eof);
                                    // Advance to next line
                                    source_char_read_result.loc.char_n = CharNumberOneBased(1);
                                    source_char_read_result.loc.line_n.inc();
                                } else {
                                    //eprintln!("advancing to next char: {:?}", &_char_or_eof);
                                    // Advance to next char
                                    source_char_read_result.loc.char_n.inc();
                                }
                            }
                        }
                        Result::Err(utf8error) => {
                            match utf8error.error_len() {
                                None => {
                                    // Need more bytes for this char
                                    need_more_bytes_for_this_char = true;
                                    // continue
                                }
                                Some(len) => {
                                    // An invalid Utf8 byte was encountered
                                    debug_assert_eq!(utf8error.valid_up_to(), 0);

                                    let char_error = CharError::InvalidUtf8(by);
                                    yield SourceCharReadResult {
                                        char_result: CharResult::CharError(char_error),
                                        ..source_char_read_result.clone()
                                    };
                                }
                            } // match
                        } // Err
                    } // match from_utf8
                } // ByteOrEof::Byte
                ByteOrEof::StdIoError(err_str) => {
                    yield SourceCharReadResult {
                        char_result: CharResult::CharError(CharError::StdIoError(err_str)),
                        ..source_char_read_result.clone()
                    };
                }
                ByteOrEof::Eof => {
                    if utf8_buf_len != 0 {
                        yield SourceCharReadResult {
                            char_result: CharResult::CharError(CharError::UnexpectedEof),
                            ..source_char_read_result.clone()
                        };
                    } else {
                        if source_char_read_result.loc.char_n != CharNumberOneBased(1) {
                            yield SourceCharReadResult {
                                char_result: CharResult::Eol,
                                ..source_char_read_result.clone()
                            };
                            source_char_read_result.loc.char_n.inc();
                        }

                        yield SourceCharReadResult {
                            char_result: CharResult::Eof,
                            ..source_char_read_result.clone()
                        };
                    };
                }
            } // match
        } // for in source_bytes
    }) // from_generator
}

#[cfg(test)]
mod test {
    #[test]
    fn test() {
        use crate::source_bytes::source_bytes;
        use crate::source_chars::{source_chars, SourceCharReadResult};

        const TEST_DATA_SUBDIR: &'static str = "source_chars";

        crate::test_util::insta_glob(TEST_DATA_SUBDIR, |file_path, bx_bufread| {
            let mut source_bytes = source_bytes(bx_bufread);
            let mut source_chars = source_chars(&mut source_bytes);

            let results = source_chars.collect::<Vec<_>>();

            insta::assert_ron_snapshot!(results);
        });
    }
}
