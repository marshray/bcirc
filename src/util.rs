// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

#![allow(dead_code)] //? TODO for development
#![allow(unused_mut)] //? TODO for development
#![allow(unused_variables)] //? TODO for development
#![allow(unused_imports)] //? TODO for development

/// Build an Option<> from a conditional and an expression.
///
/// # Examples
///
/// ```
/// # use std::ops::Range;
/// # use bcirc::maybe_some;
/// struct X(usize);
/// impl X {
///     const VALID_RANGE: Range<usize> = 1_usize .. 101;
///     fn opt_x_from_usize(val: usize) -> Option<X> {
///         maybe_some!(Self::VALID_RANGE.contains(&val), X(val))
///     }
/// }
/// assert!(X::opt_x_from_usize(100).is_some());
/// assert!(X::opt_x_from_usize(101).is_none());
/// ```
///
#[macro_export]
macro_rules! maybe_some {
    ($e:expr, $f:expr) => {
        if $e {
            std::option::Option::Some($f)
        } else {
            std::option::Option::None
        }
    };
}

/// Like `std::matches`, but allows multiple patterns.
///
/// # Examples
///
/// ```
/// # use bcirc::matches_any;
/// fn get_message(chars: &[char; 2]) -> bool {
///     matches_any!(chars,
///         &[ a, _b ] if a == 'E',
///         &[ _a, b ] if b == 'H' )
/// }
/// assert_eq!(get_message(&['E', 'F']), true);
/// assert_eq!(get_message(&['F', 'G']), false);
/// assert_eq!(get_message(&['G', 'H']), true);
/// ```
#[macro_export]
macro_rules! matches_any {
    ($ex:expr, $( $pa:pat $(if $gu:expr)? ),*) => {
        match $ex {
            $( $pa $(if $gu)? => true, )*
            _ => false
        }
    };
}

/// Build an Option<> from a match expression.
///
/// # Examples
///
/// ```
/// # use bcirc::maybe_match;
/// struct CharStruct { ch: char };
/// let char_struct = CharStruct { ch: 'F' };
/// let opt_ch: Option<char> = maybe_match!(
///     char_struct, CharStruct { ch, .. } if ch != '\0' => ch
/// );
/// assert!(opt_ch.is_some());
/// ```
#[macro_export]
macro_rules! maybe_match {
    ($scrutinee:expr, $patt:pat $( if $guard:expr )? => $some_expr:expr) => {
        match $scrutinee {
            $patt $(if $guard)? => std::option::Option::Some($some_expr),
            _ => std::option::Option::None
        }
    };
}

/// A trait for fallible conversions when no Error type is specified.
/// TODO rename MaybeFrom?
pub trait OptFrom<T>: Sized {
    fn opt_from(value: T) -> Option<Self>;

    /// Produce a `std::result::Result<T, &'static str>` with a generic error message.
    fn try_from_ok_or_err(val: T) -> Result<Self, &'static str> {
        Self::opt_from(val).ok_or("Couldn't convert val.")
    }
}

/* Alas.

error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
   = note: only traits defined in the current crate can be implemented for a type parameter

impl<T: OptFrom<U>, U> std::convert::TryFrom<U> for T
where
    T: OptFrom<U>,
{
    type Error = &'static str;

    fn try_from(val: U) -> Result<Self, Self::Error> {
        val.opt_from(val).ok_or(Err("OptFrom couldn't convert."))
    }
} */

pub(crate) const fn one_shl(ch: char) -> u128 {
    let ch = ch as u32;
    if ch < 128 {
        1_u128 << ch
    } else {
        assert!(ch < 128);
        0
    }
}

pub(crate) const fn fs_shl(n: u32, ch: char) -> u128 {
    let ch = ch as u32;
    assert!(n + ch < 128);
    ((1_u128 << n) - 1) << ch
}

pub(crate) const fn u128_ch_bit_test(u: u128, ch: char) -> bool {
    let ch = ch as u32;
    ch < 128 && ((u >> ch) & 1) != 0
}
