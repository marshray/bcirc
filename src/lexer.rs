// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

#![allow(dead_code)] //? TODO for development
#![allow(unused_mut)] //? TODO for development
#![allow(unused_variables)] //? TODO for development
#![allow(unused_imports)] //? TODO for development

use std::iter::FusedIterator;
use std::ops::RangeInclusive;

use serde::{Deserialize, Serialize};

use crate::data_repr::IntegerRepr;
use crate::source_chars::{CharError, CharLoc, CharResult, SourceCharReadResult};
use crate::util::{fs_shl, one_shl, u128_ch_bit_test};

fn is_punctuation_single_char(ch: char) -> bool {
    #[rustfmt::skip]
    const PUNCTUATION_SINGLE_CHAR: u128 =
        one_shl('[') | one_shl(']') |
        one_shl('{') | one_shl('}') |
        one_shl('(') | one_shl(')') |
        one_shl('-') | one_shl('!') |
        one_shl('^') | one_shl('&') |
        one_shl('*') | one_shl(',') |
        one_shl(';');

    u128_ch_bit_test(PUNCTUATION_SINGLE_CHAR, ch)
}

fn is_punctuation_first_char(ch: char) -> bool {
    #[rustfmt::skip]
    const PUNCTUATION_FIRST_CHAR: u128 =
        one_shl('<') | one_shl('>') |
        one_shl('=') | one_shl('&');

    u128_ch_bit_test(PUNCTUATION_FIRST_CHAR, ch)
}

const DECIMAL_DIGIT_CHAR: u128 = fs_shl(10, '0');

fn is_decimal_digit(ch: char) -> bool {
    u128_ch_bit_test(DECIMAL_DIGIT_CHAR, ch)
}

fn decimal_digit_val(ch: char) -> u32 {
    debug_assert!(is_decimal_digit(ch));

    ch as u32 - b'0' as u32
}

fn is_comment_first_char(ch: char) -> bool {
    ch == '/'
}

#[rustfmt::skip]
const IDENTIFIER_FIRST_CHAR: u128 =
    fs_shl(26, 'A') | 
    fs_shl(26, 'a') |
    one_shl('_');

fn is_identifier_first_char(ch: char) -> bool {
    u128_ch_bit_test(IDENTIFIER_FIRST_CHAR, ch)
}

fn is_identifier_subsequent_char(ch: char) -> bool {
    const IDENTIFIER_SUBSEQUENT_CHAR: u128 = IDENTIFIER_FIRST_CHAR | DECIMAL_DIGIT_CHAR;
    u128_ch_bit_test(IDENTIFIER_SUBSEQUENT_CHAR, ch)
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum Token {
    PunctuationSingleChar(char),
    Comment,
    LiteralInteger(IntegerRepr),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TokenInfo {
    pub span: [CharLoc; 2],
    pub token: Token,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum LexError {
    UnexpectedEof { char_loc_start: CharLoc },
    UnexpectedChar(char),
    UnexpectedCharResult(CharResult),
    Todo,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum LexResult {
    Token(TokenInfo),
    Eod { loc: CharLoc },
    Error { loc: CharLoc, lex_error: LexError },
}

impl LexResult {
    pub fn is_eod(&self) -> bool {
        matches!(self, &LexResult::Eod { .. })
    }
    pub fn is_error(&self) -> bool {
        matches!(self, &LexResult::Eod { .. })
    }
    pub fn is_final(&self) -> bool {
        !matches!(self, &LexResult::Token(_))
    }
}

struct Lexer<'a> {
    source_chars: &'a mut dyn Iterator<Item = SourceCharReadResult>,
    read_results: [SourceCharReadResult; 3],
    yielded_final: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(source_chars: &'a mut dyn Iterator<Item = SourceCharReadResult>) -> Self {
        let mut self_ = Self {
            source_chars,
            read_results: [
                SourceCharReadResult::default_error(),
                SourceCharReadResult::default_error(),
                SourceCharReadResult::default_error(),
            ],
            yielded_final: false,
        };
        self_.consume_char();
        self_.consume_char();
        self_
    }

    // Convenient access to the char just consumed read result
    #[inline]
    fn prev_read_result(&self) -> &SourceCharReadResult {
        &self.read_results[0]
    }

    // Convenient access to the current char read result
    #[inline]
    fn cur_read_result(&self) -> &SourceCharReadResult {
        &self.read_results[1]
    }

    // Convenient access to the next char read result
    #[inline]
    fn peek_next_read_result(&self) -> &SourceCharReadResult {
        &self.read_results[2]
    }

    /// Call this to consume a char.
    /// You don't have to call this if you're returning LexResult::is_final() (i.e., something other
    /// than a token).
    fn consume_char(&mut self) {
        eprintln!(
            "Lexer::read_next_char() consuming {:?}",
            &self.read_results[1]
        );

        self.read_results.rotate_left(1);

        eprintln!(
            "Lexer::read_next_char() cur is now {:?}",
            &self.read_results[1]
        );

        self.read_results[2] = if let Some(read_result) = self.source_chars.next() {
            read_result
        } else if self.cur_read_result().is_eof() || self.cur_read_result().is_error() {
            self.cur_read_result().clone()
        } else {
            SourceCharReadResult::default_error()
        }
    }

    /// Consumes the current char.
    /// Returns None if the new read_result is Char or Eol.
    /// Otherwise, returns an appropriate LexResult.
    ///
    /// Call this when you need to advance to the next char in order to probably use it.
    /// But if you're simply advancing past the last char you are consuming, then
    /// you can just call `consume_char()`.
    ///
    /// Actually, since all tokens end on non-chars, you probably never need this.
    ///
    // fn consume_char_expecting_char_or_eol(&mut self) -> Option<LexResult> {
    //     self.consume_char();
    //     self.lexerror_unless_cur_char_or_eol()
    // }

    /// Returns None if the cur_read_result is Char or Eol.
    /// Otherwise, returns an appropriate LexResult.
    ///
    /// Call this before you need to interpret the current read_result as a char.
    ///
    fn lexerror_unless_cur_char_or_eol(&self) -> Option<LexResult> {
        let cur_read_result = self.cur_read_result();
        if cur_read_result.is_char_or_eol() {
            None
        } else {
            Some(LexResult::Error {
                loc: cur_read_result.loc.clone(),
                lex_error: LexError::UnexpectedCharResult(cur_read_result.char_result.clone()),
            })
        }
    }

    /// Lexes the next token (or other result) from the initial state.
    fn lex_initial(&mut self) -> Option<LexResult> {
        eprintln!("Lexer::lex_initial()");

        match self.cur_read_result() {
            SourceCharReadResult {
                loc,
                char_result: CharResult::Char(ch),
            } => match *ch {
                ch if ch.is_whitespace() => {
                    self.consume_char();
                    None
                }
                ch if is_comment_first_char(ch) => self.lex_comment_first_char(),
                ch if is_punctuation_single_char(ch) => Some(self.lex_punctuation_single_char()),
                ch if is_decimal_digit(ch) => Some(self.lex_literal_digit()),

                // Every other char is unexpected
                ch => {
                    // Don't consume
                    Some(LexResult::Error {
                        loc: loc.clone(),
                        lex_error: LexError::UnexpectedChar(ch),
                    })
                }
            },
            SourceCharReadResult {
                char_result: CharResult::Eol,
                ..
            } => {
                self.consume_char();
                None
            }
            SourceCharReadResult {
                loc,
                char_result: CharResult::Eof,
            } => {
                // Don't consume
                Some(LexResult::Eod { loc: loc.clone() })
            }
            SourceCharReadResult { loc, char_result } => {
                // Don't consume
                Some(LexResult::Error {
                    loc: loc.clone(),
                    lex_error: LexError::UnexpectedCharResult(char_result.clone()),
                })
            }
        }
    }

    /// Lexes a punctuation token comprised of a single char.
    fn lex_punctuation_single_char(&mut self) -> LexResult {
        let first_char_loc = self.cur_read_result().loc.clone();

        let mut last_char_loc = first_char_loc.clone();
        let mut ch = self.cur_read_result().char_or_nul();
        assert!(is_punctuation_single_char(ch));

        self.consume_char();

        LexResult::Token(TokenInfo {
            span: [first_char_loc, last_char_loc],
            token: Token::PunctuationSingleChar(ch),
        })
    }

    fn lex_comment_first_char(&mut self) -> Option<LexResult> {
        assert!(is_comment_first_char(self.cur_read_result().char_or_nul()));
        let first_char_loc = self.cur_read_result().loc.clone();

        // Consume first '/'
        self.consume_char();

        let ch = self.cur_read_result().char_or_nul();

        if ch == '/' {
            // Consume second '/' and any following chars
            loop {
                self.consume_char();

                if !self.cur_read_result().is_char() {
                    return None;
                }
            }
        } else if ch == '*' {
            let mut state = 0usize;
            loop {
                // Consume '*' and any following chars through "*/"
                self.consume_char();

                if state == 2 {
                    return None; // Success
                }

                if !self.cur_read_result().is_char_or_eol() {
                    return None; // Some kind of problem. Let the outer loop find it.
                }

                state = match (state, self.cur_read_result().char_or_nul()) {
                    (0, '*') => 1,
                    (1, '/') => 2,
                    _ => 0,
                };
            }
        } else {
            // Don't consume it

            // Just return it as a Token::PunctuationSingleChar(ch).
            let last_char_loc = first_char_loc.clone();

            Some(LexResult::Token(TokenInfo {
                span: [first_char_loc, last_char_loc],
                token: Token::PunctuationSingleChar('/'),
            }))
        }
    }

    fn lex_literal_digit(&mut self) -> LexResult {
        assert!(is_decimal_digit(self.cur_read_result().char_or_nul()));

        let first_char_loc = self.cur_read_result().loc.clone();

        let mut last_char_loc = self.cur_read_result().loc.clone();
        let mut ch = self.cur_read_result().char_or_nul();
        let mut i: i128 = decimal_digit_val(ch).into();

        'more_digits: loop {
            self.consume_char();

            ch = self.cur_read_result().char_or_nul();
            if !is_decimal_digit(ch) {
                break 'more_digits;
            }

            i *= 10;
            i += i128::from(decimal_digit_val(ch));

            last_char_loc = self.cur_read_result().loc.clone();
        } // 'more_digits

        LexResult::Token(TokenInfo {
            span: [first_char_loc, last_char_loc],
            token: Token::LiteralInteger(IntegerRepr::I128(i)),
        })
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = LexResult;

    fn next(&mut self) -> Option<LexResult> {
        eprintln!("Lexer::next()");

        if self.yielded_final {
            return None;
        }

        loop {
            let opt_lex_result = self.lex_initial();
            if let Some(ref lex_result) = opt_lex_result {
                self.yielded_final = lex_result.is_final();
                eprintln!("Lexer::yielded_final <- {:?}", self.yielded_final);

                eprintln!("Lexer::next() yielding {:?}", &lex_result);
                return opt_lex_result;
            }
        }
    }
}

impl<'a> std::iter::FusedIterator for Lexer<'a> {}

#[cfg(test)]
mod test {
    #[test]
    fn test() {
        const TEST_DATA_SUBDIR: &str = "lexer";

        crate::test_util::insta_glob(TEST_DATA_SUBDIR, |file_path, bx_bufread| {
            let mut source_bytes = crate::source_bytes::source_bytes(bx_bufread);
            let mut source_chars = crate::source_chars::source_chars(&mut source_bytes);

            let mut tokens = crate::lexer::Lexer::new(&mut source_chars);

            let results = tokens.collect::<Vec<_>>();

            insta::assert_ron_snapshot!(results);
        });
    }
}
