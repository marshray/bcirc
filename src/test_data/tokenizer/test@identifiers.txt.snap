---
source: src/lexer.rs
input_file: src/test_data/lexer/identifiers.txt
---
[
  Token(TokenInfo(
    linechar_range: RangeInclusive(
      start: LineCharNums((1, 1)),
      end: LineCharNums((1, 1)),
    ),
    token: Identifier("a"),
  )),
  Token(TokenInfo(
    linechar_range: RangeInclusive(
      start: LineCharNums((2, 1)),
      end: LineCharNums((2, 2)),
    ),
    token: Identifier("b2"),
  )),
  Token(TokenInfo(
    linechar_range: RangeInclusive(
      start: LineCharNums((3, 1)),
      end: LineCharNums((3, 2)),
    ),
    token: Identifier("c_"),
  )),
  Token(TokenInfo(
    linechar_range: RangeInclusive(
      start: LineCharNums((4, 1)),
      end: LineCharNums((4, 3)),
    ),
    token: Identifier("d_5"),
  )),
  Eod(
    file_size: 16,
    last_line: LineNum(4),
  ),
]
